### PR TITLE
Created SharedInteractionSystem + Partial Fix #782

### LINKED ISSUE #782: InRangeUnobstructed bug
Using insideBlockerValid = true has an edge case. If the target is behind something of the collision mask but still under 1 range away, you can interact through the mask. Ex: through a diagonal line of walls. 

An example of building an apc through a diagonal wall
https://gyazo.com/bfb742e539f9f4c1e48571d5a284defc

### PR DESCRIPTION
**Pull request by SamV522**
This will partially fix #782

Still learning these PR/Commits, think I've got the hang of it moving forwards though.

Fix being able to pick up items through diagonal walls, more work required to address different entity types (Lockers are highlighted in red, but still operate, etc.)

### CODE REVIEW

File: Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        /// <returns>True if the two points are within a given range without being obstructed.</returns>
+        public bool InRangeUnobstructed(MapCoordinates coords, Vector2 otherCoords, float range = InteractionRange,
+            int collisionMask = (int)CollisionGroup.Impassable, IEntity ignoredEnt = null, bool insideBlockerValid = false)
+        {
+            bool retVal = false;
+            var dir = otherCoords - coords.Position;
+
+            if (dir.LengthSquared.Equals(0f)) return true;
+            if (range > 0f && !(dir.LengthSquared <= range * range)) return false;
+                    var ray = new CollisionRay(coords.Position, dir.Normalized, collisionMask);
```
Comment (PJB3005) [inline Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs:49]: You should really fix the indentation in this file, it's a mess to read.

File: Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                    var rayResults = _physicsManager.IntersectRay(coords.MapId, ray, dir.Length, ignoredEnt, true);
+
+                    if(!rayResults.DidHitObject || (insideBlockerValid && rayResults.DidHitObject && rayResults.Distance < 1f))
+                    {
+                        _mapManager.TryFindGridAt(coords, out var mapGrid);
+                        var srcPos = mapGrid.MapToGrid(coords);
+                        var destPos = new GridCoordinates(otherCoords, mapGrid);
+                        if (srcPos.InRange(_mapManager, destPos, InteractionRange))
+                        {
+                            retVal = true;
```
Comment (PJB3005) [inline Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs:59]: Why not just straight up `return true;` here?

Comment (SamV522) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs:59]: The return value is set to true so the Method has a return value on all codepaths, see line 63.

Currently, carrying the return value from start to finish is a catch all for the IF statements not having an "else" case, which could be used in future to do more if it's needed.

File: Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs (hunk 1 above)
Comment (SamV522) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs]: New commit to address the indendation, I think this is how you wanted it?

File: Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs (hunk 2 above)
Comment (PJB3005) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs]: You can just put a `return false;` at the end. It makes the code way easier to follow:

```cs
public bool InRangeUnobstructed(MapCoordinates coords, Vector2 otherCoords, float range = InteractionRange,
    int collisionMask = (int)CollisionGroup.Impassable, IEntity ignoredEnt = null, bool insideBlockerValid = false)
{
    var dir = otherCoords - coords.Position;

    if (dir.LengthSquared.Equals(0f)) return true;
    if (range > 0f && !(dir.LengthSquared <= range * range)) return false;

    var ray = new CollisionRay(coords.Position, dir.Normalized, collisionMask);
    var rayResults = _physicsManager.IntersectRay(coords.MapId, ray, dir.Length, ignoredEnt, true);
    if(!rayResults.DidHitObject || (insideBlockerValid && rayResults.DidHitObject && rayResults.Distance < 1f))
    {
        _mapManager.TryFindGridAt(coords, out var mapGrid);
        var srcPos = mapGrid.MapToGrid(coords);
        var destPos = new GridCoordinates(otherCoords, mapGrid);
        if (srcPos.InRange(_mapManager, destPos, InteractionRange))
        {
            return true;
        }
    }

    return false;
}```

File: Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs (hunk 1 above)
Comment (PJB3005) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs]: Yep.

**Pull request merged**